### PR TITLE
feat: Support send acp/sudt to any address.

### PR DESCRIPTION
**Pull request by yanguoyu**
1. acp/sudt -> secp256k1 address
     
     i. Cheque lock cell

https://user-images.githubusercontent.com/12881040/170648962-aa45a093-2c4c-4148-b574-8aa23771c03b.mov

     ii. secp256k1/sudt cell

https://user-images.githubusercontent.com/12881040/170651043-cacab398-f5ff-4cee-8c31-9e6bd8fb6b4b.mov

2. acp/sudt -> acp/sudt
      
     i. Send to exist acp/sudt cell
      
https://user-images.githubusercontent.com/12881040/170655405-a8ae1133-9e2d-41c6-b0cb-2dc9b06a95d3.mov

      
      ii. create a acp/sudt cell

https://user-images.githubusercontent.com/12881040/170657768-fc2dc906-4cb8-4bdf-a6b5-3e41666697ac.mov

3. acp/sudt -> unknow address

https://user-images.githubusercontent.com/12881040/170658526-d8f64385-1ec5-4672-870e-899d7b166e0e.mov

### CODE REVIEW

**Comment by Keith-CY** [comment]
image
Use radio instead of checkbox if only one option is allowed

image
We should not imply user to make a decision, simply disable the button

image
We should not check it by default, simply disable the button

**Comment by yanguoyu** [comment]
> image Use radio instead of checkbox if only one option is allowed
> 
> image We should not imply user to make a decision, simply disable the button
> 
> image We should not check it by default, simply disable the button

Generally, radio buttons can't be canceled and can only be switched. If there is only one option, should we just show tips to tell the user how to send sudt? And no need to show radio button?

**Comment by Keith-CY** [comment]
> > image Use radio instead of checkbox if only one option is allowed
> > image We should not imply user to make a decision, simply disable the button
> > image We should not check it by default, simply disable the button
> 
> Generally, radio buttons can't be canceled and can only be switched. If there is only one option, should we just show tips to tell the user how to send sudt? And no need to show radio button?

What would it be like to show tips to tell the user how to send sudt

**Comment by yanguoyu** [comment]
> > > image Use radio instead of checkbox if only one option is allowed
> > > image We should not imply user to make a decision, simply disable the button
> > > image We should not check it by default, simply disable the button
> > 
> > 
> > Generally, radio buttons can't be canceled and can only be switched. If there is only one option, should we just show tips to tell the user how to send sudt? And no need to show radio button?
> 
> What would it be like to show tips to tell the user how to send sudt

Like this:
image
Delete checkbox, similar like send cheque before, only show tips.

**Comment by Keith-CY** [comment]
> > > > image Use radio instead of checkbox if only one option is allowed
> > > > image We should not imply user to make a decision, simply disable the button
> > > > image We should not check it by default, simply disable the button
> > > 
> > > 
> > > Generally, radio buttons can't be canceled and can only be switched. If there is only one option, should we just show tips to tell the user how to send sudt? And no need to show radio button?
> > 
> > 
> > What would it be like to show tips to tell the user how to send sudt
> 
> Like this: image Delete checkbox, similar like send cheque before, only show tips.

Ah.. I this there may be a misunderstanding

I mean the first case, there're to options, but only one is allowed to pick, we should use radio, because radio means you have to pick one but only one.

The other cases, only one option to select or unselect, can be a checkbox, because checkbox means you have to accept it to the next step.

**Comment by yanguoyu** [comment]
> > > > > image Use radio instead of checkbox if only one option is allowed
> > > > > image We should not imply user to make a decision, simply disable the button
> > > > > image We should not check it by default, simply disable the button
> > > > 
> > > > 
> > > > Generally, radio buttons can't be canceled and can only be switched. If there is only one option, should we just show tips to tell the user how to send sudt? And no need to show radio button?
> > > 
> > > 
> > > What would it be like to show tips to tell the user how to send sudt
> > 
> > 
> > Like this: image Delete checkbox, similar like send cheque before, only show tips.
> 
> Ah.. I think there may be a misunderstanding
> 
> I mean the first case, there're to options, but only one is allowed to pick, we should use radio, because radio means you have to pick one but only one.
> 
> The other cases, only one option to select or unselect, can be a checkbox, because checkbox means you have to accept it to the next step.

Means when options' length is greater than 1, use radio buttons? Otherwise, use checkbox?

**Comment by Keith-CY** [comment]
> > > > > > image Use radio instead of checkbox if only one option is allowed
> > > > > > image We should not imply user to make a decision, simply disable the button
> > > > > > image We should not check it by default, simply disable the button
> > > > > 
> > > > > 
> > > > > Generally, radio buttons can't be canceled and can only be switched. If there is only one option, should we just show tips to tell the user how to send sudt? And no need to show radio button?
> > > > 
> > > > 
> > > > What would it be like to show tips to tell the user how to send sudt
> > > 
> > > 
> > > Like this: image Delete checkbox, similar like send cheque before, only show tips.
> > 
> > 
> > Ah.. I think there may be a misunderstanding
> > I mean the first case, there're to options, but only one is allowed to pick, we should use radio, because radio means you have to pick one but only one.
> > The other cases, only one option to select or unselect, can be a checkbox, because checkbox means you have to accept it to the next step.
> 
> Means when options' length is greater than 1, use radio buttons? Otherwise, use checkbox?

Almost there, we use radio button when 1 is mandatory while more than 1 are available. And we use checkboxes if it's multiple selectable or only 1 to agree/disagree.

**Comment by yanguoyu** [comment]
Use radio:

image

Remove default select:

image

**Review by qiweiii — APPROVED** [review]

File: packages/neuron-ui/src/components/SUDTSend/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  acp = 'acp',
+  unknow = 'unknow',
+}
+
+export function useAddressLockType(address: string, isMainnet: boolean) {
+  return useMemo(() => {
+    if (isSecp256k1Address(address)) {
+      return AddressLockType.secp256
+    }
+    if (isAnyoneCanPayAddress(address, isMainnet)) {
```
Comment (Keith-CY) [inline packages/neuron-ui/src/components/SUDTSend/hooks.ts:28]: `PW Lock ACP Address` is missing here.

Ref: https://github.com/Magickbase/ckb-wallet-and-explorer/wiki/Neuron-Functionalities
Ref: https://github.com/nervosnetwork/neuron/blob/develop/packages/neuron-wallet/.env#L30

File: packages/neuron-wallet/src/controllers/anyone-can-pay.ts [hunk 2]
```diff
@@ -54,11 +46,16 @@ export default class AnyoneCanPayController {
     }
   }
 
-  public async generateSendAllTx(params: GenerateAnyoneCanPayAllTxParams): Promise<Controller.Response<Transaction>> {
-    return this.generateTx({
-      ...params,
-      amount: 'all'
-    })
+  public async getSudtCellExtraCkb(
```
[the remark this answers is not in the document]
Comment (yanguoyu) [reply packages/neuron-wallet/src/controllers/anyone-can-pay.ts]: rename to `getHoldSudtCellCapacity`.

**Review by Keith-CY — APPROVED** [review]

**Comment by Keith-CY** [comment]
https://user-images.githubusercontent.com/7271329/171573483-148d0cd2-76bb-4b0a-b4d3-b46d6758ae7f.mov

Balance is enough but an error is thrown.

**Comment by yanguoyu** [comment]
> When the balance is insufficient, Migrate to a new sUDT Asset Account, error box is displayed (as shown in the right image below) Expected: When click Migrate button, will info be prompted at the top (as shown in the left image below)?
> 
> Wallet c1 - Neuron

fixed by https://github.com/nervosnetwork/neuron/pull/2400/commits/18c55b49dda2dc48b517b1dfa60bac59d1d1f85d

**Comment by Keith-CY** [comment]
How is it going @Cedar67

**Comment by Cedar67** [comment]
The following scenarios were verified and passed. 

&#x2705;  1 - acp/sudt -> secp256k1 address: Cheque lock cell or secp256k1/sudt cell.
&#x2705;  2 - acp/sudt -> acp/sudt: 
- > Send to exist acp/sudt cell.
- > create a acp/sudt cell.

&#x2705;  3 - acp/sudt -> unknow address: create a unknow/sudt cell.
&#x2705;  4 - acp/ckb -> secp256k1 address: Create a secp256k1 lock cell.
&#x2705;  5 - acp/ckb -> acp/ckb address: 
- > Send to exist address
- > Send to unknow address: throw error

Today I'll also test some exceptions in detail and verify [18c55b4](https://github.com/nervosnetwork/neuron/pull/2400#:~:text=fixed%20by-,18c55b4,-Keith%2DCY%20mentioned). If everything goes well, I will approve this PR today.

**Review by Cedar67 — APPROVED** [review]

**Pull request merged**